packages/gatsby/src/redux/persist.ts

```diff
@@ -1,12 +1,175 @@
+import path from "path"
 import v8 from "v8"
-import { readFileSync, writeFileSync } from "fs-extra"
-import { ICachedReduxState } from "./types"
+import {
+  existsSync,
+  mkdtempSync,
+  readFileSync,
+  removeSync,
+  renameSync,
+  writeFileSync,
+} from "fs-extra"
+import { IReduxNode, ICachedReduxState } from "./types"
+import { sync as globSync } from "glob"
+import report from "gatsby-cli/lib/reporter"
 
-const file = (): string => `${process.cwd()}/.cache/redux.state`
+const getLegacyCacheFile = (): string =>
+  // TODO: remove this legacy stuff in v3 (fairly benign change but still)
+  // This is a function for the case that somebody does a process.chdir (#19800)
+  path.join(process.cwd(), `.cache/redux.state`)
 
-export const readFromCache = (): ICachedReduxState =>
-  v8.deserialize(readFileSync(file()))
+const getReduxCacheFolder = (): string =>
+  // This is a function for the case that somebody does a process.chdir (#19800)
+  path.join(process.cwd(), `.cache/redux`)
 
-export const writeToCache = (contents: ICachedReduxState): void => {
-  writeFileSync(file(), v8.serialize(contents))
+function reduxRestFile(dir: string): string {
```

Comment:
I find restFile such an ugly term XD What about making it more sexy with "global" :joy:

```suggestion
function reduxGlobalFile(dir: string): string {
```

Reply:
Also consider "shared" (as in the file is shared for multiple redux substates), maybe?